File: modules/text-scanner/ios/TextScannerModule.swift
import ExpoModulesCore
import Vision
import UIKit

/// On-device text recognition backed by Apple's Vision framework.
///
/// Vision is used rather than ML Kit because Google's ML Kit CocoaPods ship no
/// arm64 iOS-simulator slice and force `EXCLUDED_ARCHS[sdk=iphonesimulator*]`,
/// which makes the app unbuildable for the simulator on Apple Silicon. Vision
/// is first-party, adds no binary weight, and never leaves the device.
public class TextScannerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TextScanner")

    AsyncFunction("scanImage") { (uri: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let result = try Self.scan(uri: uri)
          promise.resolve(result)
        } catch {
          promise.reject("ERR_TEXT_SCAN", error.localizedDescription)
        }
      }
    }
  }

  private enum ScanError: LocalizedError {
    case unreadableImage

    var errorDescription: String? {
      switch self {
      case .unreadableImage: return "The image could not be read for text recognition."
      }
    }
  }

  private static func scan(uri: String) throws -> [String: Any] {
    let url = URL(string: uri) ?? URL(fileURLWithPath: uri)

    guard let data = try? Data(contentsOf: url),
          let image = UIImage(data: data),
          let cgImage = image.cgImage else {
      throw ScanError.unreadableImage
    }

    let width = CGFloat(cgImage.width)
    let height = CGFloat(cgImage.height)

    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    // Sensitive values are codes, not prose; language correction rewrites card
    // and IBAN digit groups into dictionary words and destroys the match.
    request.usesLanguageCorrection = false

    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    try handler.perform([request])

    let observations = request.results ?? []
    var lines: [[String: Any]] = []

    for observation in observations {
      guard let candidate = observation.topCandidates(1).first else { continue }
      let text = candidate.string
      if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { continue }

      // Vision reports a normalised box with origin at the bottom-left; the
      // rest of the pipeline works in top-left pixel space.
      let lineBox = Self.pixelRect(observation.boundingBox, width: width, height: height)

      // Per-word boxes come from the recognised-text range API rather than
      // being interpolated, so a redaction can cover one field instead of the
      // whole line.
      //
      // The search start advances past each match. Without it, `range(of:)`
      // returns the first occurrence every time, so in "4242 4242 4242 4242"
      // all four groups resolve to the first group's box and the redaction
      // covers only the leading digits.
      var words: [[String: Any]] = []
      var searchStart = text.startIndex
      for word in text.split(separator: " ", omittingEmptySubsequences: true) {
        guard let range = text.range(of: String(word), range: searchStart..<text.endIndex) else {
          continue
        }
        searchStart = range.upperBound

        guard let wordBox = try? candidate.boundingBox(for: range) else { continue }
        let rect = Self.pixelRect(wordBox.boundingBox, width: width, height: height)
        words.append([
          "text": String(word),
          "x": rect.origin.x,
          "y": rect.origin.y,
          "width": rect.size.width,
          "height": rect.size.height,
        ])
      }

      lines.append([
        "text": text,
        "x": lineBox.origin.x,
        "y": lineBox.origin.y,
        "width": lineBox.size.width,
        "height": lineBox.size.height,
        "words": words,
      ])
    }

    return [
      "imageWidth": Int(width),
      "imageHeight": Int(height),
      "lines": lines,
    ]
  }

  /// Converts a normalised, bottom-left-origin Vision box to top-left pixels.
  private static func pixelRect(_ box: CGRect, width: CGFloat, height: CGFloat) -> CGRect {
    CGRect(
      x: box.minX * width,
      y: (1 - box.maxY) * height,
      width: box.width * width,
      height: box.height * height
    )
  }
}

File: modules/text-scanner/android/src/main/java/expo/modules/textscanner/TextScannerModule.kt
package expo.modules.textscanner

import android.graphics.BitmapFactory
import android.graphics.Rect
import android.net.Uri
import androidx.exifinterface.media.ExifInterface
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * On-device text recognition backed by ML Kit's bundled Latin recogniser.
 *
 * The bundled variant is used rather than the Play-Services-delivered one so
 * the first scan works with no network and no model download, which is what the
 * app's privacy claim requires.
 */
class TextScannerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TextScanner")

    AsyncFunction("scanImage") { uri: String, promise: Promise ->
      try {
        scan(uri, promise)
      } catch (error: Exception) {
        promise.reject("ERR_TEXT_SCAN", error.message ?: "Text recognition failed.", error)
      }
    }
  }

  private fun scan(uri: String, promise: Promise) {
    val context = appContext.reactContext
      ?: return promise.reject("ERR_TEXT_SCAN", "No React context available.", null)

    val parsed = Uri.parse(uri)
    val image = InputImage.fromFilePath(context, parsed)

    // InputImage applies the EXIF rotation, so the recogniser's boxes are in
    // upright space; the reported dimensions must match that, not the raw file.
    val rotated = image.rotationDegrees == 90 || image.rotationDegrees == 270
    val width = if (rotated) image.height else image.width
    val height = if (rotated) image.width else image.height

    TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
      .process(image)
      .addOnSuccessListener { result ->
        val lines = mutableListOf<Map<String, Any>>()

        for (block in result.textBlocks) {
          for (line in block.lines) {
            val lineBox = line.boundingBox ?: continue
            if (line.text.isBlank()) continue

            val words = line.elements.mapNotNull { element ->
              element.boundingBox?.let { box -> boxMap(element.text, box) }
            }

            lines.add(
              mapOf(
                "text" to line.text,
                "x" to lineBox.left,
                "y" to lineBox.top,
                "width" to lineBox.width(),
                "height" to lineBox.height(),
                "words" to words,
              )
            )
          }
        }

        promise.resolve(
          mapOf(
            "imageWidth" to width,
            "imageHeight" to height,
            "lines" to lines,
          )
        )
      }
      .addOnFailureListener { error ->
        promise.reject("ERR_TEXT_SCAN", error.message ?: "Text recognition failed.", error)
      }
  }

  private fun boxMap(text: String, box: Rect): Map<String, Any> = mapOf(
    "text" to text,
    "x" to box.left,
    "y" to box.top,
    "width" to box.width(),
    "height" to box.height(),
  )
}
